once_test.py

```diff
@@ -167,25 +327,83 @@ def yielding_iterator():
         self.assertEqual(list(yielding_iterator()), [1, 2, 3])
         self.assertEqual(list(yielding_iterator()), [1, 2, 3])
 
+    def test_failing_generator(self):
+        counter = Counter()
+
+        @once.once
+        def sample_failing_fn():
+            yield counter.get_incremented()
+            result = counter.get_incremented()
+            yield result
+            if result == 2:
+                raise ValueError("expected failure after 2.")
+
+        # Both of these calls should return the same results.
+        call1 = sample_failing_fn()
+        call2 = sample_failing_fn()
+        self.assertEqual(next(call1), 1)
+        self.assertEqual(next(call2), 1)
+        self.assertEqual(next(call1), 2)
+        self.assertEqual(next(call2), 2)
+        with self.assertRaises(ValueError):
+            next(call1)
+        with self.assertRaises(ValueError):
+            next(call2)
+        # These next 2 calls should succeed.
+        call3 = sample_failing_fn()
+        call4 = sample_failing_fn()
+        self.assertEqual(list(call3), [3, 4])
+        self.assertEqual(list(call4), [3, 4])
+        # Subsequent calls should return the original value.
+        self.assertEqual(list(sample_failing_fn()), [3, 4])
+        self.assertEqual(list(sample_failing_fn()), [3, 4])
+
     def test_iterator_parallel_execution(self):
         counter = Counter()
 
+        @execute_with_barrier(n_workers=_N_WORKERS)
```

Comment:
Maybe worth a comment that this requires executor.map to be run over an integer multiple of _N_WORKERS so we can reuse the same barrier

Reply:
done